src/Plutarch/Extra/Time.hs

```diff
@@ -0,0 +1,159 @@
+     which is more efficient in usage.
+
+     @since 3.3.0
+-}
+data PCurrentTime (s :: S) = PCurrentTime
+    { lowerBound :: Term s PPOSIXTime
+    , upperBound :: Term s PPOSIXTime
+    }
+    deriving stock
+        ( -- | @since 3.3.0
+          Generic
+        )
+    deriving anyclass
+        ( -- | @since 3.3.0
+          PlutusType
+        , -- | @since 3.3.0
+          PEq
+        )
+
+-- | @since 3.3.0
+instance DerivePlutusType PCurrentTime where
+    type DPTStrat _ = PlutusTypeScott
+
+{- | Get the current time, given a 'PPOSIXTimeRange'.
+
+     If it's impossible to get a fully-bounded time, (e.g. either end of the 'PPOSIXTimeRange' is
+     an infinity) then we return nothing.
+
+     @since 3.3.0
+-}
+pcurrentTime ::
+    forall (s :: S).
+    Term
+        s
+        ( PPOSIXTimeRange
+            :--> PMaybe PCurrentTime
+        )
+pcurrentTime = phoistAcyclic $
+    plam $ \iv -> unTermCont $ do
+        PInterval iv' <- pmatchC iv
+        ivf <- pletAllC iv'
+        PLowerBound lb <- pmatchC ivf.from
+        PUpperBound ub <- pmatchC ivf.to
+
+        let getBound = phoistAcyclic $
+                plam $
+                    flip pletAll $ \f ->
+                        pif
+                            f._1
+                            ( pmatch f._0 $ \case
+                                PFinite (pfromData . (pfield @"_0" #) -> d) -> pjust # d
+                                _ ->
+                                    ptrace
+                                        "pcurrentTime: time range should be bounded"
+                                        pnothing
+                            )
+                            (ptrace "pcurrentTime: time range should be inclusive" pnothing)
+
+            lb' = getBound # lb
+            ub' = getBound # ub
+
+            mkTime = phoistAcyclic $ plam $ pcon .* PCurrentTime
+        pure $ pliftA2 # mkTime # lb' # ub'
+
+{- | Calculate current time providing the @validaRange@ field, which typically
+      comes from 'PTxInfo'.
+
+     @since 3.3.0
+-}
+currentTime ::
+    forall r (s :: S).
+    (HasField "validRange" r (Term s PPOSIXTimeRange)) =>
+    r ->
+    Term s (PMaybe PCurrentTime)
+currentTime x = pcurrentTime # x.validRange
+
+{- | Calculate current time, and error out with the given message if we can't
+      get one.
+
+     @since 3.3.0
+-}
+passertCurrentTime ::
+    forall (s :: S).
+    Term
+        s
+        ( PString
+            :--> PPOSIXTimeRange
+            :--> PCurrentTime
+        )
+passertCurrentTime = phoistAcyclic $
+    plam $
+        \msg iv -> passertPJust # msg #$ pcurrentTime # iv
+
+{- | Retutn true if a `PPOSIXTime` is in the current time range.
+
+     @since 3.3.0
+-}
+pisWithinCurrentTime ::
+    forall (s :: S).
+    Term
+        s
+        ( PPOSIXTime
+            :--> PCurrentTime
+            :--> PBool
+        )
+pisWithinCurrentTime = phoistAcyclic $
```

Comment:
It can be `plam $ \t -> pisCurrentTimeWithin # t` I believe.

Reply:
I don't think it can. `pisWithinCurrentTime` tell you a point of time is in the current time range, while  `pisCurrentTimeWithin` is supposed to yield true if current time range is in the given time range - the first argument of `pisCurrentTimeWithin` is actually the lower bound of the said time range.